byterun/alloc.c

```diff
@@ -53,14 +51,116 @@ CAMLexport value caml_alloc (mlsize_t wosize, tag_t tag)
   return result;
 }
 
+static inline void enter_gc_preserving_vals(mlsize_t wosize, value* vals)
+{
+  CAMLparam0();
+  /* Copy the values to be preserved to a different array.
+     The original vals array never escapes, generating better code in
+     the fast path. */
+  CAMLlocalN(vals_copy, wosize);
+  mlsize_t i;
+  for (i = 0; i < wosize; i++) vals_copy[i] = vals[i];
+  caml_handle_gc_interrupt();
+  for (i = 0; i < wosize; i++) vals[i] = vals_copy[i];
+  CAMLreturn0;
+}
+
+static inline value do_alloc_small(mlsize_t wosize, tag_t tag, value* vals)
+{
+  value v;
+  mlsize_t i;
+  Assert (tag < 256);
+  Alloc_small(v, wosize, tag,
+      { enter_gc_preserving_vals(wosize, vals); });
+  for (i = 0; i < wosize; i++) {
+    Op_val(v)[i] = vals[i];
+  }
+  return v;
+}
+
+
+CAMLexport value caml_alloc_1 (tag_t tag, value a)
+{
+  value v[1] = {a};
+  return do_alloc_small(1, tag, v);
+}
+
+CAMLexport value caml_alloc_2 (tag_t tag, value a, value b)
+{
+  value v[2] = {a, b};
+  return do_alloc_small(2, tag, v);
+}
+
+CAMLexport value caml_alloc_3 (tag_t tag, value a, value b, value c)
+{
+  value v[3] = {a, b, c};
+  return do_alloc_small(3, tag, v);
+}
+
+CAMLexport value caml_alloc_4 (tag_t tag, value a, value b, value c, value d)
+{
+  value v[4] = {a, b, c, d};
+  return do_alloc_small(4, tag, v);
+}
+
+CAMLexport value caml_alloc_5 (tag_t tag, value a, value b, value c, value d,
+                               value e)
+{
+  value v[5] = {a, b, c, d, e};
+  return do_alloc_small(5, tag, v);
+}
+
+CAMLexport value caml_alloc_6 (tag_t tag, value a, value b, value c, value d,
+                               value e, value f)
+{
+  value v[6] = {a, b, c, d, e, f};
```

Comment:
Can this and similar allocations in `caml_alloc_N` be `CAMLlocalN`? Since the very next thing we do is `do_alloc_small()` -> `enter_gc_preserving_calls()` which copies the values from this array to `CAMLlocalN` structure.